hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java

```diff
@@ -40,96 +44,345 @@
+   * @param node DatanodeUsageInfo to be searched for.
+   * @return true if the specified node is present in listToSearch, otherwise
+   * false.
+   */
+  private boolean containsNode(
+      List<DatanodeUsageInfo> listToSearch, DatanodeUsageInfo node) {
+    int index = 0;
+    Comparator<DatanodeUsageInfo> comparator =
+        DatanodeUsageInfo.getMostUsedByRemainingRatio();
+
+    if (comparator.compare(listToSearch.get(0),
+        listToSearch.get(listToSearch.size() - 1)) < 0) {
+      index =
+          Collections.binarySearch(listToSearch, node, comparator.reversed());
+    } else {
+      index = Collections.binarySearch(listToSearch, node, comparator);
+    }
+    return index >= 0 && listToSearch.get(index).equals(node);
+  }
+
+  /**
+   * Calculates the number of used bytes given capacity and utilization ratio.
+   *
+   * @param nodeCapacity capacity of the node.
+   * @param utilizationRatio used space by capacity ratio of the node.
+   * @return number of bytes
+   */
+  private double ratioToBytes(Long nodeCapacity, double utilizationRatio) {
+    return nodeCapacity * utilizationRatio;
+  }
+
+  /**
+   * Calculates the average datanode utilization for the specified nodes.
+   * Utilization is used space divided by capacity.
+   *
+   * @param nodes List of DatanodeUsageInfo to find the average utilization for
+   * @return Average utilization value
+   * @throws ArithmeticException Division by zero
+   */
+  private double calculateAvgUtilization(List<DatanodeUsageInfo> nodes)
+      throws ArithmeticException {
     SCMNodeStat aggregatedStats = new SCMNodeStat(
         0, 0, 0);
     for (DatanodeUsageInfo node : nodes) {
       aggregatedStats.add(node.getScmNodeStat());
     }
-    return aggregatedStats.getScmUsed().get().doubleValue() /
-        aggregatedStats.getCapacity().get().doubleValue();
+    clusterCapacity = aggregatedStats.getCapacity().get();
+    clusterUsed = aggregatedStats.getScmUsed().get();
+    clusterRemaining = aggregatedStats.getRemaining().get();
+
+    try {
+      return clusterUsed / (double) clusterCapacity;
+    } catch (ArithmeticException e) {
+      LOG.warn("Division by zero while calculating average utilization of the" +
+          " cluster in ContainerBalancer.");
+      throw e;
+    }
+  }
+
+  /**
+   * Calculates the utilization, that is used space divided by capacity, for
+   * the given datanodeUsageInfo.
+   *
+   * @param datanodeUsageInfo DatanodeUsageInfo to calculate utilization for
+   * @return Utilization value
+   */
+  private double calculateUtilization(DatanodeUsageInfo datanodeUsageInfo) {
+    SCMNodeStat stat = datanodeUsageInfo.getScmNodeStat();
+
+    return stat.getScmUsed().get().doubleValue() /
+        stat.getCapacity().get().doubleValue();
   }
 
+  /**
+   * Stops ContainerBalancer.
+   */
   public void stop() {
     LOG.info("Stopping Container Balancer...");
-    balancerRunning = false;
+    balancerRunning.set(false);
```

Comment:
NIT. remove line 319, one info line is sufficient here, since no actual work needs to be done here.